File: cpp/gps.cpp
#include <algorithm>
#include <iostream>
#include <vector>

#include "gps.hpp"
#include "lane.hpp"
#include "map.hpp"
#include "road.hpp"
#include "roundabout.hpp"

using namespace std;


//  Constructors/destructor
GPS::GPS() : algorithm(A_STAR) {}

GPS::~GPS() {}


//  Returns approximatively the time needed to reach another node
//  A naive heuristics : straight line distance
float GPS::heuristic_weight(const Roundabout* origin, const Roundabout* destination) const {
    float distance = abs(*origin - *destination);
    float speed    = Road::DEFAULT.max_speed();

    return distance/*/speed*/;
}


//  Performs an A* pathfinding to get a path from origin to destination
//  Returns a list of integers, representing the roads to follow
vector<Roundabout*> GPS::find_path(Roundabout* origin, Roundabout* destination) const { 
    //   Before we start, let's check we need to do something
    if (origin == destination || !abs(origin - destination))
        return vector<Roundabout*>();
    

    //   A* algorithm
    if (algorithm == A_STAR) {
        //   Initialization
        vector<Roundabout*> open_list;
        vector<Roundabout*> closed_list;

        open_list.push_back(origin);
        
        origin->set_g_cost(0);
        origin->set_h_cost(heuristic_weight(origin, destination));

        //  Mainloop
        Roundabout*     current_parent(0);
        unsigned short  i(0), slots_count(0);
        bool            found(false);

        while (open_list.size() && !found) {
            current_parent = open_list.back();
            open_list.pop_back();
            slots_count = current_parent->slots_count();
            i = 0;
            while (i < slots_count) {
                Roundabout* child   = current_parent->slot(i)->lane()->to();
                Road*   road    = current_parent->slot(i)->lane()->road();
                float current_g_cost = current_parent->g_cost() + road->weight();
                float current_h_cost = heuristic_weight(child, destination);
                
                //  No leaving lane => next
                if (child == current_parent) {i++; continue;}
                
                //  Path found
                if (child == destination)   {found = true; child->set_nearest_parent(current_parent); break;}
                

                //   Not yet in the closed list neither in the open list
                //   TODO: binary_search didn't work properly ; try to find why
                bool in_open_list(false);
                bool in_closed_list(false);
                unsigned short k(0);
                while (k < open_list.size()) {
                    if (open_list[k] == child)  {in_open_list = true;}
                    k++;
                }
                k = 0;
                while (k < closed_list.size()) {
                    if (closed_list[k] == child)  {in_closed_list = true;}
                    k++;
                }

                if (!in_open_list &&  !in_closed_list) {
                    //   Store data for this child
                    child->set_destination(destination);
                    child->set_nearest_parent(current_parent);
                    child->set_g_cost(current_g_cost);
                    child->set_h_cost(current_h_cost);

                    //   Add the node to the open list, keeping the order (the first node has the smallest F-cost)
                    unsigned int j(0);
                    while (j < open_list.size() && open_list[j]->f_cost() > child->f_cost())    {j++;}

                    open_list.insert(open_list.begin() + j, child);
                }


                //   Already in the open list => check if this path is a better one than the already known path
                else if (in_open_list && !in_closed_list) {
                    //   This path is shorter than the existing one => store it
                    if (current_g_cost < child->g_cost()) {
                        child->set_nearest_parent(current_parent);
                        child->set_g_cost(current_g_cost);
                        child->set_h_cost(current_h_cost);

                        //   Do not forget to keep the open-list sorted
                        remove(open_list.begin(), open_list.end(), child);

                        unsigned short j(0);
                        while (j < open_list.size() && open_list[j]->f_cost() > child->f_cost())    {j++;}
    
                        open_list.insert(open_list.begin() + j, child);
                    }
                }

                i++;
            }

            //   The parent node, around which we look, is always the first node of the "open" list
            //   Transfert the parent node to the "closed" list
            closed_list.push_back(current_parent);
        }


        //   Save the path if it exists.
        if (found) {
            Roundabout*         current_node = destination;
            vector<Roundabout*> path;

            while (current_node != origin) {
                path.insert(path.begin(), current_node);

                current_node = current_node->nearest_parent();
                if (!current_node)
                    cerr << "ERROR (in GPS::find_path()): ill-formed parent list, a node has no parent.";
            }
            
            return path;
        }

        return vector<Roundabout*>();
    }


    //   TODO : Dijkstra
    else if (algorithm == DIJKSTRA)
        return vector<Roundabout*>();


    //   Unknown algorithm
    return vector<Roundabout*>();
}

File: python/gps.py
﻿# -*- coding: utf-8 -*-
"""
File        :   gps.py
Description :   provides pathfinding information
"""

from constants  import *
import lib

class Gps():
    """
    
    """

    def __init__(self, new_algorithm = A_STAR):
        """

        """

        self.algorithm      = new_algorithm
        
        self.closed_list    = []
        self.open_list      = []

        self.path_length    = 0
        self.real_cost      = 0
        self.g_cost         = {}
        self.h_cost         = {}
        self.f_cost         = {}
        self.path           = []
        self.start_time     = 0
        
    def _build_path(self):
        """
        Given a list of nodes, returns a list of integers accounting for the road number at a roundabout.
        """

        path = []
        
        for i in range(len(self.path) - 1):
            current_node    = self.path[i]
            next_node       = self.path[i + 1]
            
            key_list = [i for i in range(len(current_node.leaving_roads)) if current_node.leaving_roads[i].end == next_node]
            
            if len(key_list) == 0:
                raise Exception('ERROR (in gps._build_path()) : there is no route.')
                
            path.append(key_list[0])
            
        return path
        
    def _heuristic_weight(self, origin, destination):
        """
        Returns an evaluation of the time needed to go from a node to another.
        """

        #   A naive heuristics : straight line distance
        distance = abs(origin.position - destination.position)
        speed    = ROAD_DEFAULT_MAX_SPEED
        
        return distance/speed
        
    def find_path(self, origin, destination, max_time = 1):
        """
        Performs an A* pathfinding to get a path from origin to destination.
        Returns a list of integers, representing the roads to take.
        
            origin (Roundabout)
            destination   (Roundabout)
            max_time   (float)     : maximum time allowed for the pathfinding algorithm (in seconds)
        """
        
        #   Before we start, let's check we need to do something
        if origin == destination or self._heuristic_weight(origin, destination) == 0:
            return None
        
        #   Add the starting point to the "open" list
        self.open_list.append(origin)
        self.g_cost[origin] = 0
        self.h_cost[origin] = self.f_cost[origin] = self._heuristic_weight(origin, destination)
        
        self.start_time = lib.clock()
        nearest_parent  = {}
        self.path       = PATH_INEXISTENT
        
        #while (lib.clock() - self.start_time) < max_time:
        while len(self.open_list):
            #   The "parent" node, around which we look, is always the first node of the "open" list
            #   This node is transferred to the "closed" list
            current_parent = self.open_list[0]
            self.closed_list.append(current_parent)
            del self.open_list[0]

            #   The "parent" node is the destination : the path has been found.
            if current_parent == destination:
                self.path = PATH_FOUND
                break

            #   Set the first element of the open list as the one that has the smallest F-cost
            for (i, node) in enumerate(self.open_list):
                if self.f_cost[self.open_list[0]] > self.f_cost[node]:
                    (self.open_list[i], self.open_list[0]) = (self.open_list[0], node)
            
            #   Check the adjacent nodes
            children = [road.end for road in current_parent.leaving_roads]
            
            for child in children:
                #   Not already in the closed list neither in the open list
                if not (child in self.closed_list) and not (child in self.open_list):
                    #   Compute its G-cost, H-cost and F-cost
                    self.g_cost[child] = self.g_cost[current_parent] + road.weight
                    self.h_cost[child] = self._heuristic_weight(child, destination)
                    self.f_cost[child] = self.g_cost[child] + self.h_cost[child]
                    
                    nearest_parent[child] = current_parent
                    
                    #   Add the node to the open list, keeping the order (the first node has the smallest F-cost)
                    if len(self.open_list) and (self.f_cost[self.open_list[0]] > self.f_cost[child]):
                        self.open_list.insert(0, child)
                    else:
                        self.open_list.append(child)

                #   Already in the open list : check to see if this path is a better one than the currently known path
                elif child in self.open_list:
                    #   Compute the G-cost of this possible new path
                    current_g_cost = self.g_cost[current_parent] + road.weight
                        
                    #   This path is shorter (lower G-cost) : store this path as default to reach this node
                    if current_g_cost < self.g_cost[child]:
                        #   Set this path as the shortest path to reach this node
                        nearest_parent[child]    = current_parent
                        self.g_cost[child]       = current_g_cost
                        self.f_cost[child]       = self.g_cost[current_parent] + self.h_cost[child]   # Do not forget to update the F-cost !
                        
                        #   Check if the open list is still in the right order
                        if self.f_cost[self.open_list[0]] > self.f_cost[child]:
                            i = self.open_list.index(child)
                            (self.open_list[0], self.open_list[i]) = (self.open_list[i], self.open_list[0])

        #   Save the path if it exists.
        if self.path == PATH_FOUND:
            
            current_node        = destination
            self.path           = []
            self.path_length = 0
            
            while current_node != origin:
                self.path.insert(0, current_node)
                if current_node in nearest_parent:
                    current_node = nearest_parent[current_node]
                else:
                    raise Exception('ERROR (in gps.find_path()): ill-formed parent list, a node has no parent.')
                    
                self.path_length += 1
            return self._build_path()

        return None